packages/api-bindings/src/apollo/cache/utils/cursorBasedPagination.ts

```diff
@@ -27,34 +23,71 @@ export function cursorBasedPagination<T extends CursorBasedPagination>(
 
       // items that are not in the cache anymore (for .e.g deleted publication)
       const danglingItems = items.filter((item) => !canRead(item));
-      const readRes = {
+      const readRes: SafeReadonly<TResult> = {
         ...existing,
         items,
         pageInfo: {
           ...pageInfo,
+          beforeCount: existing.pageInfo.beforeCount ?? 0,
+          moreAfter: existing.pageInfo.moreAfter ?? existing.pageInfo.next !== null,
           // reduce total count by excluding dangling items so it won't cause a new page query
           // after item was removed from the cache (for .e.g deleted publication)
           totalCount:
             pageInfo.totalCount !== null ? pageInfo.totalCount - danglingItems.length : null,
         },
-      } as SafeReadonly<T>;
+      };
 
       return readRes;
     },
 
-    merge(existing: Readonly<T> | undefined, incoming: SafeReadonly<T>) {
-      if (!existing) {
-        return incoming;
+    merge(
+      existing: Readonly<TResult> | undefined,
+      incoming: SafeReadonly<TResult>,
+      { variables = {} }: FieldFunctionOptions,
+    ) {
+      if (!isCursor(variables.cursor) || !existing) {
+        return {
+          ...incoming,
+          pageInfo: {
+            ...incoming.pageInfo, // future-proofing in case we add more fields to pageInfo
+            beforeCount: 0, // assume there is no newer results than the provided
+            moreAfter: incoming.pageInfo.next !== null,
+          },
+        };
       }
 
       const existingItems = existing.items;
       const incomingItems = incoming.items;
 
-      return {
-        ...incoming,
-        items: existingItems.concat(incomingItems),
-        pageInfo: incoming.pageInfo,
-      } as SafeReadonly<T>;
+      if (variables.cursor === existing.pageInfo.prev) {
+        return {
+          ...incoming,
+          items: incomingItems.concat(existingItems),
+          pageInfo: {
+            ...incoming.pageInfo, // future-proofing in case we add more fields to pageInfo
+            beforeCount: incoming.pageInfo.beforeCount,
+            moreAfter: existing.pageInfo.next !== null,
+            next: existing.pageInfo.next,
+            prev: incoming.pageInfo.prev ?? existing.pageInfo.prev,
+          },
+        };
+      }
+
+      if (variables.cursor === existing.pageInfo.next) {
+        return {
+          ...incoming,
+          items: existingItems.concat(incomingItems),
+          pageInfo: {
+            ...incoming.pageInfo, // future-proofing in case we add more fields to pageInfo
+            beforeCount: incoming.pageInfo.beforeCount,
+            moreAfter: incoming.pageInfo.next !== null,
+            next: incoming.pageInfo.next ?? existing.pageInfo.next,
+            prev: existing.pageInfo.prev,
+          },
+        };
+      }
+
+      throw new InvariantError('Unable to merge incoming cursor-based pagination result');
```

Comment:
nit, we have also a `never` helper that we use across the mono repo. Not strong as both are ok

Reply:
I checked `never` throws a simple `Error`.

Did you mean `invariant` helper?

Reply:
`invariant` would require you to pass some value and seeing `invariant(false, "...")` IMO is confusing.
But looking at the `never` implementation i think it should return `InvariantError` as i looked quickly through the all usages of the helper and it's always used as an invariant that should never happen.